core/jobs/batch_jobs/opportunity_management_jobs.py

```diff
@@ -43,19 +45,179 @@
     from mypy_imports import datastore_services
     from mypy_imports import exp_models
     from mypy_imports import opportunity_models
+    from mypy_imports import question_models
+    from mypy_imports import skill_models
     from mypy_imports import story_models
     from mypy_imports import topic_models
 
 (
     exp_models, opportunity_models, story_models,
-    topic_models
+    topic_models, skill_models, question_models
 ) = models.Registry.import_models([
     models.NAMES.exploration, models.NAMES.opportunity, models.NAMES.story,
-    models.NAMES.topic
+    models.NAMES.topic, models.NAMES.skill, models.NAMES.question
 ])
 datastore_services = models.Registry.import_datastore_services()
 
 
+class DeleteSkillOpportunityModelJob(base_jobs.JobBase):
+    """Job that deletes SkillOpportunityModels."""
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        """Returns a PCollection of 'SUCCESS' or 'FAILURE' results from
+        deleting SkillOpportunityModel.
+
+        Returns:
+            PCollection. A PCollection of 'SUCCESS' or 'FAILURE' results from
+            deleting SkillOpportunityModel.
+        """
+        skill_opportunity_model = (
+            self.pipeline
+            | 'Get all non-deleted skill models' >> ndb_io.GetModels(
+                opportunity_models.SkillOpportunityModel.get_all(
+                    include_deleted=False))
+        )
+
+        unused_delete_result = (
+            skill_opportunity_model
+            | beam.Map(lambda model: model.key)
+            | 'Delete all models' >> ndb_io.DeleteModels()
+        )
+
+        return (
+            skill_opportunity_model
+            | 'Create job run result' >> (
+                job_result_transforms.CountObjectsToJobRunResult())
+        )
+
+
+class GenerateSkillOpportunityModelJob(base_jobs.JobBase):
+    """Job for regenerating SkillOpportunityModel.
+
+    NOTE: The DeleteSkillOpportunityModelJob must be run before this
+    job.
+    """
+
+    @staticmethod
+    def _count_unique_question_ids(
+        question_skill_link_models: list[
+                question_models.QuestionSkillLinkModel
+            ]
+        ) -> int:
+        """Counts the number of unique question ids.
+
+        Args:
+            question_skill_link_models: list(QuestionSkillLinkModel).
+                Array of QuestionSkillLinkModels.
+
+        Returns:
+            int. The number of unique question ids.
+        """
+
+        question_ids = [link.question_id for link in question_skill_link_models]
+        return len(set(question_ids))
+
+    @staticmethod
+    def _create_skill_opportunity_model(
+        skills_and_links: dict
+    ) -> result.Result[
```

Comment:
Modify this to accept the skills and links as separate args and then down in the `run` method unpack the dictionary using lambda (`lambda object: _create_skill_opportunity_model(object['skill'], object['question_skill_links'])`), this should improve the readability of this function.

Reply:
Good point, thanks! Done